### PR TITLE
MQTT: Automation module

### PR DESCRIPTION
**Pull request by davidgraeff**
Includes a Trigger and Action handler
to replace the old MQTT Actions binding.

Signed-off-by: David Graeff <david.graeff@web.de>

### CODE REVIEW

**Comment by htreu** [comment]
Your daily Travis report:

```
handlerInitializeAfterSubscribe(org.eclipse.smarthome.binding.mqtt.internal.MQTTMultiBrokerTopicDiscoveryTest)  Time elapsed: 0.007 sec
Results :
Failed tests: 
  BrokerHandlerTest.handlerInit:110 
callback.statusUpdated(
    thing,
    <Capturing argument>
);
Wanted 3 times:
-> at org.eclipse.smarthome.binding.mqtt.handler.BrokerHandlerTest.handlerInit(BrokerHandlerTest.java:110)
But was 4 times. Undesired invocation:
-> at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:399)
```

**Comment by htreu** [comment]
This actually seems to be unrelated. Maybe the `MQTTMultiBrokerTopicDiscoveryTest` is not as stable as it should be?

**Comment by htreu** [comment]
We have a WIP framework update which addresses these binding specific automation actions. See https://github.com/openhab/openhab2-addons/pull/2926#issuecomment-424977138 for reference. It will ease the development a lot.
@davidgraeff we should wait for @triller-telekom to comment here and sync the development.

**Comment by triller-telekom** [comment]
Thanks for pinging me. As you already mentioned these ruleActions from me are still WIP.

@davidgraeff: How "soon" do you "need" them? :-)

**Comment by triller-telekom** [comment]
Thanks for pinging me. As you already mentioned these ruleActions from me are still WIP.

@davidgraeff: How "soon" do you "need" them? :-)

**Comment by davidgraeff** [comment]
Not many people are using the new rule engine yet, but as soon as that is the case (experimental flag removed, and introductional Blog article or similar) MQTT should work with it

**Comment by kaikreuzer** [comment]
> As you already mentioned these ruleActions from me are still WIP.

@triller-telekom Could you maybe create a WIP PR with work? Would be great, because we could show the idea where this is heading and also make it visible, how much it has advanced resp. how much is still left to do. And it would give people the chance to comment early, if they see design flaws or important use cases not being covered.

**Comment by triller-telekom** [comment]
here we go: https://github.com/eclipse/smarthome/pull/6398

**Comment by davidgraeff** [comment]
This module would be necessary anyway, because of the triggers and conditions it provides.

I do like idea of creating/attaching action handlers to thing handlers though and might indeed be added by me at some point.

The current action handler adds 80 lines of code and requires the user to configure it with a system broker connection. That is still a useful action handler in my opinion, so the WIP PR has no affect on this PR.

File: features/karaf/esh-core/src/main/feature/feature.xml [hunk 1]
```diff
@@ -212,7 +212,7 @@
     <feature dependency="true">esh-automation-module-script</feature>
     <bundle>mvn:org.eclipse.smarthome.automation/org.eclipse.smarthome.automation.module.media/${project.version}</bundle>
   </feature>
-
+
```
Comment (htreu) [inline features/karaf/esh-core/src/main/feature/feature.xml]: unrelated

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    @Override
+    public Collection<ModuleType> getModuleTypes(Locale locale) {
+        cachedLocale = locale;
+        return Stream.of(getActionType(locale), getConditionType(locale), getTriggerType(locale))
+                .collect(Collectors.toList());
+    }
+
+    private ModuleType getActionType(Locale locale) {
+        final ActionType actionType = new ActionType(PublishActionHandler.MODULE_TYPE_ID, getConfigActionDesc(locale), //
+                "Publish MQTT message", "Publishes a message to a MQTT topic", //
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: My intuition also tells me to use "an MQTT", but may popular sites (hivemq, wikipedia) say "a MQTT". Google also finds "an MQTT".

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Override
+    public void topicVanished(ThingUID thingUID, MqttBrokerConnection connection, @NonNull String topic) {
+        receivedPayload = "";
+    }
+
+    @NonNullByDefault({}) // TODO Until the ConditionHandler interface is fixed
+    @Override
+    public boolean isSatisfied(Map<String, Object> context) {
+        return expectedPayload.equals(receivedPayload);
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java]: Because MQTT allows retained topic values (which translates into a state for a topic), the condition also makes perfect sense to me. The description might need to be changed to reflect the use case. The condition of course does not make sense for non-retained topics.

The trigger should fire on every new published message on a given topic, there are people in the community that require this behavior. But I do see the point. Another trigger could be introduced that triggers only on a specific message on a given topic.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Constants
+ *
+ * @author David Graeff - Initial contribution
+ *
+ */
+@NonNullByDefault()
+public class MQTTModuleConstants {
+    public static final String TOPIC_NAME_TYPE = "mqtt.topic.name";
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java]: I wasn't sure about the scope, there is not much docu yet. Other modules (I have looked at the timing and example one) also have a prefix AFAIK, but could be wrong. Other modules definitely did the "dot" syntax, but I'm fine with came case.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+Export-Package: org.eclipse.smarthome.automation.module.mqtt.handler
+Import-Package: 
+ com.google.common.collect,
+ org.apache.commons.lang,
+ org.eclipse.jdt.annotation;resolution:=optional,
+ org.eclipse.smarthome.automation,
+ org.eclipse.smarthome.automation.handler,
+ org.eclipse.smarthome.automation.module.mqtt.handler,
+ org.eclipse.smarthome.automation.type,
+ org.eclipse.smarthome.binding.mqtt.discovery,
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF]: I tried to stay independent, that is the reason why the Action handler only allows system brokers to publish messages to. On the other hand, all user configured broker connections will not be usable. I guess this specific automation module just requires its subsystem binding.

What I can imagine to do is making that dependency optional. Either only system broker connections (MQTTService ones) are used or all available if the binding is installed.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: Don't use Google to figure out english grammar ;-)
Go here: https://www.quora.com/Is-it-correct-to-use-an-before-M-when-its-in-its-initial-form-When-I-say-an-MRI-it-sounds-correct-but-when-I-write-it-that-way-I-feel-like-I-violated-the-a-an-rule

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java]: > Because MQTT allows retained topic values

But this implementation does not actively "request" those retained values from the broker or does it? What would be the use case where one would need them instead of relying on an item being filled with a value and using the existing item condition?

> Another trigger could be introduced that triggers only on a specific message on a given topic.

No! The existing MQTT trigger should simply have an optional parameter - this is exactly how it is done in the ItemStateTrigger, so we should be in line with that.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF]: > all user configured broker connections will not be usable

Could you briefly explain why?

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java]: The module types use "core.itemStateTrigger" and such stuff, but I haven't seen any input/output ids using that syntax - actually, dots in those most likely will break the framework, if you look at the feature implemented in https://github.com/eclipse/smarthome/pull/6229.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java]: Agreed. So an optional configuration for a trigger message will be added.

The implementation does not actively subscribe to retained values, that is correct. It's not possible actually. You can filter messages later on, if you want to, because messages have the retained flag set by the broker. That is not possible in the current MQTTBrokerConnection implementation though (you are usually not interested if a topic value is retained or not).

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF]: The broker connections that are created in the MQTT binding are not inserted into the MQTTService. They exist as Bridges only.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java]: Inputs/Outputs can be processed and manipulated by other module conditions/actions as well, can't they? So prefixing them isn't a bad idea at all, if the name is as generic as "topicName"?

[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleConstants.java]: Outputs are references as `<moduleId>.<outputId>` so prefixing then IS a bad idea.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java]: Sounds like a good reason then to remove the condition - at least for now.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF]: You're right, I didn't remember that detail in the architecture :-/.

>  the Action handler only allows system brokers to publish messages to

Does this make sense then? Wouldn't people expect to be able to use the MQTT bridges that they have configured to be usable as well for actions?

> I guess this specific automation module just requires its subsystem binding.

Which makes it to be an extension of that binding, just like the generic mqtt thing extension. I'm wondering if it wouldn't make sense to move it to there, i.e. change the bundle to become `org.eclipse.smarthome.binding.mqtt.automation` (we should come up with a package name convention together with #6229 then). As the system brokers are made available as bridges as well, we could then imho simply say that ALL modules require a bridge id as their configuration, which would simplify the options.

> What I can imagine to do is making that dependency optional. Either only system broker connections (MQTTService ones) are used or all available if the binding is installed.

In the past I sometimes thought optional dependencies would be a neat idea as well, but in almost all situations it turned out to be a bad decision, especially as DS components do not go well with that at all.

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java]: What about this use-case:

**WHEN** *mqtt_topic_updated*("sensors/window1") with *mqtt_message*("open")
**IF** *mqtt_topic*("alarmpanel/state") *equals* *mqtt_message*("ARMED")
**THEN** ...

MQTT topics are not necessarily bound to an item.

[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishedMessageConditionHandler.java]: > " not necessarily"

Correct. But in your example, the "alarmpanel state" probably would make a very good item as well. And my big fear is people will start mixing stuff as in your example and end up with parts of their system state representation in items and other parts in the mqtt broker's retained events. That definitely should not be our design goal. I see MQTT as an "extension to the ESH event bus" and the item registry is meant to be the place, where event values are persisted (just as it has been [documented 8 years ago ;-)](https://github.com/openhab/openhab1-addons/wiki#item-repository)).

File: bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ org.eclipse.smarthome.config.core,
+ org.eclipse.smarthome.core.audio,
+ org.eclipse.smarthome.core.common.registry,
+ org.eclipse.smarthome.core.thing,
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF]: I can't, the "ThingUID" type is required.

[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/automation/org.eclipse.smarthome.automation.module.mqtt/META-INF/MANIFEST.MF]: Yes, this is linked to [this discussion](https://github.com/eclipse/smarthome/pull/6378#discussion_r228442101) - if we had concluded that it should not be an extension of the binding, we would have had to remove this dependency as well. As we decided differently, my comment here is outdated.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishActionHandler.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        String mqttbroker;
+    }
+
+    private final Logger logger = LoggerFactory.getLogger(PublishActionHandler.class);
+    private final MQTTBrokerConnectionDiscoveryService mqttService;
+    // A weak reference cache for the broker connection
+    private WeakReference<@Nullable MqttBrokerConnection> connection = new WeakReference<>(null);
+    private final Config config;
+
+    public PublishActionHandler(Action module, MQTTBrokerConnectionDiscoveryService mqttService) {
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishActionHandler.java]: This should be handed in a `BrokerHandler` and nothing else.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/MqttBrokerHandlerFactory.java [hunk 8]
```diff
@@ -46,14 +51,19 @@
  * @author David Graeff - Initial contribution
  */
 @NonNullByDefault
-@Component(immediate = true, service = { ThingHandlerFactory.class,
-        MQTTTopicDiscoveryService.class }, configurationPid = "MqttBrokerHandlerFactory")
-public class MqttBrokerHandlerFactory extends BaseThingHandlerFactory implements MQTTTopicDiscoveryService {
+@Component(immediate = true, service = { ThingHandlerFactory.class, MQTTTopicDiscoveryService.class,
+        MQTTBrokerConnectionDiscoveryService.class }, configurationPid = "MqttBrokerHandlerFactory")
+public class MqttBrokerHandlerFactory extends BaseThingHandlerFactory
+        implements MQTTTopicDiscoveryService, MQTTBrokerConnectionDiscoveryService {
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/MqttBrokerHandlerFactory.java]: This is imho all VERY misleading. Neither `MQTTTopicDiscoveryService` nor `MQTTBrokerConnectionDiscoveryService` are ESH `DiscoveryService`s - I'd therefore suggest to use a different naming than "DiscoveryService" and "DiscoveryParticipant" - why not e.g. just have `MQTTBrokerConnectionListener`s?
It also feels wrong that the handler factory implements those interfaces at all - from what I see, this can be completely removed.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/MqttBrokerHandlerFactory.java [hunk 9]
```diff
@@ -46,14 +51,19 @@
  * @author David Graeff - Initial contribution
  */
 @NonNullByDefault
-@Component(immediate = true, service = { ThingHandlerFactory.class,
-        MQTTTopicDiscoveryService.class }, configurationPid = "MqttBrokerHandlerFactory")
-public class MqttBrokerHandlerFactory extends BaseThingHandlerFactory implements MQTTTopicDiscoveryService {
+@Component(immediate = true, service = { ThingHandlerFactory.class, MQTTTopicDiscoveryService.class,
+        MQTTBrokerConnectionDiscoveryService.class }, configurationPid = "MqttBrokerHandlerFactory")
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/MqttBrokerHandlerFactory.java]: configurationPids of handler factories are always of the syntax `binding.<bindingid>`, so here probably best something like `binding.mqttbroker`.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        return actionType;
+    }
+
+    private ModuleType getTriggerType(Locale locale) {
+        return new TriggerType(PublishedMessageTriggerHandler.MODULE_TYPE_ID, getConfigTriggerDesc(locale),
+                "MQTT message published", "Triggers whenever the given MQTT topic has a new value", //
+                null, Visibility.VISIBLE, getOutputTriggerDesc(locale));
+    }
+
+    private @Nullable List<Output> getOutputTriggerDesc(Locale locale) {
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: broker & topicName are both static configuration and thus should not be an output. The only "dynamic" information is imho the topicValue, which qualifies as an output.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+                "Broker connection", "The broker connection", null, null, null);
+        return Stream.of(topicName, topicValue, brokerConnection).collect(Collectors.toList());
+    }
+
+    private @Nullable List<Input> getInputActionDesc(Locale locale) {
+        Input topicName = new Input(MQTTModuleConstants.INOUT_TOPIC_NAME, MQTTModuleConstants.TYPE_TOPIC, //
+                "Topic", "A topic that was triggered from", null, false, null, null);
+        Input topicValue = new Input(MQTTModuleConstants.INOUT_TOPIC_VALUE, MQTTModuleConstants.TYPE_STRING, //
+                "Topic", "The received topic value if triggered from a topic", null, false, null, null);
+        Input brokerConnection = new Input(MQTTModuleConstants.INOUT_BROKER_CONNECTION,
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: The broker is a configuration parameter of the action, it must not be additionally be an input.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameter.Type;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameterBuilder;
+import org.eclipse.smarthome.config.core.ParameterOption;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+import org.eclipse.smarthome.io.transport.mqtt.MqttBrokerConnection;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * Lists the MQTT system brokers for the user to choose from.
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: It lists all broker connections now, not just the MQTTService ones, true.
The rest of the provider is static and should go into a json file at some time, I guess. There are no guidelines yet though.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishActionHandler.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+            // First try the weak-reference cache
+            brokerConnection = connection.get();
+            // Fetch from the discovery service if not in the cache
+            if (brokerConnection == null) {
+                brokerConnection = mqttService.getAllBrokerConnections().get(mqttbroker);
+                connection = new WeakReference<>(brokerConnection);
+            }
+        }
+
+        // If this action was triggered by the MQTT Trigger Handler, we know a broker connection already.
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishActionHandler.java]: That's nasty. Make the broker a mandatory configuration parameter of the action and always only use that one.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishActionHandler.java (hunk 7 above)
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/handler/PublishActionHandler.java]: I see a pitfall. Assuming that a user changes his MQTT broker. He would either change the bridge configuration or remove the broker bridge and create a new one.

All his rule actions associated with that bridge would be gone then. If I keep action handlers and mqtt broker bridge handlers separate, the user notices that his MQTT actions are broken and he would only need to reassociate them with a broker.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        return Stream.of(topicName, topicValue, brokerConnection).collect(Collectors.toList());
+    }
+
+    private List<ConfigDescriptionParameter> getConfigActionDesc(Locale locale) {
+        ConfigDescriptionParameter paramBroker = ConfigDescriptionParameterBuilder
+                .create(MQTTModuleConstants.CFG_BROKER, Type.TEXT).withRequired(true).withLabel("Broker")
+                .withDescription("The broker to publish to").withOptions(getBrokerConnectionIDs())
+                .withLimitToOptions(true).build();
+        ConfigDescriptionParameter paramMessage = ConfigDescriptionParameterBuilder
+                .create(MQTTModuleConstants.CFG_MESSAGE, Type.TEXT).withRequired(true).withLabel("Message")
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: The message (topicValue) is already an input of the action, so it should not be a configuration parameter as well.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/MqttBrokerHandlerFactory.java (hunk 8 above)
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt/src/main/java/org/eclipse/smarthome/binding/mqtt/internal/MqttBrokerHandlerFactory.java]: Only the mqtt binding factory knows about all broker connections. The factory would be involved in any architecture (except if I iterate over the Thing registry, like I did before) or do I miss something?

The `MQTTTopicDiscoveryService` and `MQTTBrokerConnectionDiscoveryService` can be renamed, of course. We struggled last time to find a proper name.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java (hunk 11 above)
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: If a rule got triggered by the MQTT trigger, that exact broker connection can be used for the MQTT action and I see this as valid use case.
A boolean configuration "Trigger connection overrides configured broker" is missing to make this explicit. Wdyt?

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: The parameter should actually ideally be a UID of the broker bridge - and we could then add "Thing" as context of the according parameter with a filter for mqtt bridge types and the UI will automatically figure out the available options.

But my main point here was that this is the JavaDoc of the `MQTTModuleTypeProvider`, so the description does not match at all the primary purpose of that class.

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java (hunk 14 above)
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: I'm still confused about configuration parameters and inputs. Where can I specify inputs in PaperUI to test my rule actions?

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java (hunk 10 above)
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: For "broker": That depends on what you think about my suggestion below.
For "topicName": The user can specify a wildcard topic "sensors/+/config" for a trigger. The "topicName" is the actual topic that was received like "sensors/abc/config".

File: extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (davidgraeff) [reply extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java]: Oh I didn't know that "Thing context with filter" is possible. That indeed makes the implementation simpler.

**Comment by kaikreuzer** [comment]
@davidgraeff Let us briefly take a step back - looking at our discussions, I had a couple of déjà-vûs as I had most of that already discussed with @triller-telekom in the context of #6398.

> I absolutely agree that we actually do not need #6398 for that, so we can clearly move on.

Now that we do NOT use system brokers, but fully depend on MQTT bridges of the binding, I have to 100% revert my statement above - I think we are now just doing exactly the same as #6398 and instead of inventing it all over again, it might make sense to actually base the MQTT action on #6398 as this will give us the chance to see what might still need to be lacking or have a need to be improved. I understand that it is a bit frustrating to throw away the work of the current PR, but after all you said yourself that it is merely 80 lines of code ;-) - I hope you do agree that it'd be worthwhile to take the chance and make MQTT the official reference of how to implement actions for a binding.

Assuming we can go that route this would still leave us with the MQTT trigger. Discussing this with @triller-telekom and @htreu, we came to the conclusion that this is actually the wrong approach - we would not want every binding to add its own very specific trigger types. What has already been introduced for exactly those use cases are the Trigger channels on Things and we already have the "ChannelTrigger" module type in place in the automation engine. So for MQTT, it imho would make a lot of sense to mark the bridges to be extensible by an "mqtt message trigger channel", which would take a topic name (and optionally a topic value) as a configuration. The user could than add any number of such trigger channels to his bridge and using those in rules would work out of the box, not only for the new rule engine, but even for the DSL rules.

Wdyt, are you in to follow this path?

**Comment by davidgraeff** [comment]
Makes sense
